add some patches (bidi and vi mode runtime)

nimaaskarian opened the pull request:
- add [bidi](https://tools.suckless.org/dmenu/patches/bidi/) patch (but with some improvements for it to work with prompts and the tsv patch)
- add my own vi mode runtime starting mode patch (allows you to pass 0 or non-zero numbers to `-vi` argument to define the starting mode of vi_mode in the runtime)

nimaaskarian committed:
feat: add bidi patch

nimaaskarian committed:
fix: make bidi patch work with tsv patch

nimaaskarian committed:
feat: add vi mode runtime starting mode patch

nimaaskarian committed:
fix: fix argument reading problem in vi runtime mode patch

bakkeby commented:
It is always better to raise pull requests for individual patches separately - otherwise you have that the merging of one patch depends on the approval of the other, etc.

I can add the fribidi patch, but I think the base patch is not particularly good and will result in compatibility issues with a range of other patches, like you have already noticed with prompts and the TSV patch. I have an alternative implementation in another project so let's try with that first.

bakkeby committed:
Adding bidi patch ref. #40

Should in principle be compatible with all patches.

The bidi patch does not apply when pango is enabled,
because pango handles RTL on its own.

bakkeby committed:
bidi: apply fribidi after calculating static ellipsis and invalid widths

bakkeby commented:
The VI mode runtime starting mode patch proposal I think does not make that much sense. It has a range of issues.

First off in the config we have two variables `vi_mode` and `start_mode`.

The `vi_mode` controls whether the functionality is available or not, as in if it is set to 0 then you can't enable the functionality using a keybinding. As such to enable the functionality you would have to pass the `-vi` command line argument to enable it.

The `start_mode` is another setting that controls whether dmenu starts in VI (insert) mode or as normal when the `-vi` command line argument is passed. Otherwise this setting is not used, as in if `vi_mode` is set to 1 and you start dmenu without the command line argument then it will start in normal mode.

That is the basis of the original patch and it seems a bit overcomplicated as it is.

Now the VI mode runtime starting mode patch proposal changes the above logic to something like this:

```diff
 	#if VI_MODE_PATCH
 	} else if (!strcmp(argv[i], "-vi")) {
+		int next_arg_num = atoi(argv[++i]);
+		if (next_arg_num) {
+			start_mode = vi_mode;
+		}
-		vi_mode = 1;
 		using_vi_mode = start_mode;
 		global_esc.ksym = XK_Escape;
 		global_esc.state = 0;
 	#endif // VI_MODE_PATCH
```

It makes it hard to reason about, but the gist of it is that:
   - if `vi_mode` is set to 0 in the config, then the functionality can never be activated even when passing the `-vi` command line argument
   - if `vi_mode` is set to 1 then
      - if the following argument is 0 then we leave `start_mode` as it is set in the config
      - if the following argument is 1 then we set `start_mode` to be `vi_mode`, which is 1
   - thus if `start_mode` is set to 1 in the config then the argument that follows `-vi` has no effect
   - and if `-vi` is the last argument then `argv[++i]` may result in undefined behaviour or segmentation fault

What I may propose instead is to make the argument to `-vi` optional such that:
   - `-vi` means the normal behaviour and
   - `-vi 0` means enable the functionality and set start mode to insert mode and
   - `-vi 1` means enable the functionality and set start mode to normal mode

Something like this:

```diff
 		#if VI_MODE_PATCH
 		} else if (!strcmp(argv[i], "-vi")) {
+			if (i + 1 < argc) {
+				if (!strcmp(argv[i+1], "0")) {
+					start_mode = 0;
+					i++;
+				} else if (!strcmp(argv[i+1], "1")) {
+					start_mode = 1;
+					i++;
+				}
+			}
 			vi_mode = 1;
 			using_vi_mode = start_mode;
 			global_esc.ksym = XK_Escape;
 			global_esc.state = 0;
 		#endif // VI_MODE_PATCH
```

Would you be fine with that? We wouldn't need a separate patch option for this.

nimaaskarian commented:
yeah sorry for cramping multiple patches in one PR.

the vi mode improvement sounds cool. as dmenu doesn't really take positional arguments this shouldn't result in a conflict in other args and the optional number

bakkeby committed:
vi mode: add optional positional argument for the -vi command line option ref. #40

bakkeby committed:
Bidi + vi_mode - removing some redundant code

bakkeby committed:
Merge branch 'master' into master
